domains/client/relayer/src/lib.rs

```diff
@@ -229,200 +229,132 @@ where
-            let id = (msg.channel_id, msg.nonce);
-            let should_relay = match api.should_relay_inbox_message_response(best_hash, msg.dst_chain_id, id) {
-                Ok(valid) => valid,
-                Err(err) => {
-                    tracing::error!(
-                        target: LOG_TARGET,
-                        ?err,
-                        "Failed to fetch validity of inbox message response {id:?} for domain {0:?}",
-                        msg.dst_chain_id
-                    );
-                    false
-                }
-            };
+        };
 
-            let channel_storage = ChannelStorage::new(msg.dst_chain_id);
-            if should_relay && let Some(dst_channel_state) = channel_storage.get_channel_state_for(
-                &**consensus_client,
-                msg.src_chain_id,
-                msg.channel_id,
-            )
+        let channel_storage = ChannelStorage::new(msg.dst_chain_id);
+        if should_relay
+            && let Some(dst_channel_state) = channel_storage
+                .get_channel_state_for(&**consensus_client, msg.src_chain_id, msg.channel_id)
                 .ok()
-                .flatten() && let Some(dst_chain_outbox_response_nonce) = dst_channel_state.latest_response_received_message_nonce{
-                // relay inbox response if the dst_chain did not execute is already
-                let relay_message = msg.nonce > dst_chain_outbox_response_nonce;
-                if !relay_message{
-                    log::debug!(
-                        "Skipping message relay from {:?} to {:?}",
-                        msg.src_chain_id,
-                        msg.dst_chain_id,
-                    );
-                }
-                relay_message
-            } else { should_relay }
-        });
-
-        Ok(msgs)
-    }
-
-    pub(crate) fn submit_messages_from_consensus_chain(
-        consensus_chain_client: &Arc<Client>,
-        finalized_block: (NumberFor<Block>, Block::Hash),
-        gossip_message_sink: &GossipMessageSink,
-    ) -> Result<(), Error>
-    where
-        Client::Api: MmrApi<Block, H256, NumberFor<Block>>
-            + RelayerApi<Block, NumberFor<Block>, NumberFor<Block>, Block::Hash>,
-    {
-        // since the `finalized_block_number - 1` block MMR leaf is included in `finalized_block_number`,
-        // we process that block instead while using the finalized block number to generate the MMR
-        // proof `finalized_block_number -1`
-        let block_number_to_process = match finalized_block.0.checked_sub(&One::one()) {
-            None => return Ok(()),
-            Some(number) => number,
+                .flatten()
+        {
+            // if this message should relay,
+            // check if the dst_chain inbox nonce is more than message nonce,
+            // if so, skip relaying since message is already executed on dst_chain
+            let relay_message = msg.nonce >= dst_channel_state.next_inbox_nonce;
```

Comment:
Why is it okay to simply drop messages?

Reply:
The `dst_channel_state` is updated by a `ChannelUpdate` message, which contains a storage proof of the channel state from the dst chain and the corresponding ER from the consensus chain, if the nonce of the dst chain is greater than `msg.nonce` meaning this message is already executed on the dst chain, so we safely skip this message but the message is still keep in the state so if the ER is later proven fraudulent the relayer will retry to relay the message.

Reply:
I meant the other way around: `msg.nonce == dst_channel_state.next_inbox_nonce` is what you should probably be checking, if it is larger than there were nonces missed, which either means missed messages or some unexpected bugs, none of which seem to be good things to have.

Reply:
In the dst chain, the message is handled by the order of the nonce, one by one, if the last handles msg nonce is 1 then the dst chain will only handle msg with nonce 2. Then the channel state `next_inbox_nonce`/`latest_response_received_message_nonce ` of the dst chain will be updated.

The check here is performed at the relayer of the src chain to filter out the message that has already been relayed and executed by the dst chain (while the order of message handling is taken care of by the dst chain).

Reply:
I'm still a little bit confused. What is the main purpose of these checks then if things will be ignored/sorted on the other side anyway?